core/src/main/java/org/apache/iceberg/BaseTableScan.java

```diff
@@ -123,8 +124,14 @@ public TableScan useSnapshot(long scanSnapshotId) {
         "Cannot override snapshot, already set to id=%s", context.snapshotId());
     Preconditions.checkArgument(ops.current().snapshot(scanSnapshotId) != null,
         "Cannot find snapshot with ID %s", scanSnapshotId);
-    return newRefinedScan(
-        ops, table, schema, context.useSnapshotId(scanSnapshotId));
+    if (this instanceof DataTableScan) {
+      Schema snapshotSchema = SnapshotUtil.schemaFor(table, scanSnapshotId);
+      return newRefinedScan(
+          ops, table, snapshotSchema, context.useSnapshotId(scanSnapshotId));
+    } else {
+      return newRefinedScan(
+          ops, table, schema, context.useSnapshotId(scanSnapshotId));
```

Comment:
Looks like the implementation of `SnapshotUtil.snapshotIdAsOfTime` was copied from the `asOfTime` method below, but the method below wasn't updated to use the utility. Can you fix that? I think you can change it to this:

```java
  @Override
  public TableScan asOfTime(long timestampMillis) {
    Preconditions.checkArgument(context.snapshotId() == null,
        "Cannot override snapshot, already set to id=%s", context.snapshotId());

    return useSnapshot(SnapshotUtil.snapshotIdAsOfTime(table(), timestampMillis));
  }
```

Reply:
You're right. I'll have `BaseTableScan#asOfTime` call `SnapshotUtil.snapshotIdAsOfTime` to avoid duplicating the logic.